feat: Introduce a reusable MarkDownEditor with improved UX

lizziemac opened the pull request:
# Description

- **Fixed UI height:** The height of the Markdown UI has been adjusted to prevent it from hiding the collection tabs when the documentation is very long.
- **Improved save feedback:** A toast now pops up when the content is saved by either blurring, hitting `cmd+s`, or by clicking the `Save` button.
- **Added placeholder text:** Placeholder text has been added to guide users to write documentation of their collection!
- **Adjusted double-click behavior:** Double-clicking no longer takes you into edit mode, making it easier to copy/paste from the docs. Instead, you can quadruple click. This allows users to use the native "double-click" and "triple-click" functionality of text boxes. If this doesn't go over well, reverting is fine.
- **Enhanced button design:** The button has been redesigned and repositioned for a nicer experience.
- **Hover enables a nice background color**

Resolves #2032 

## Demos

Video of switching from placeholder text view/preview to edit mode

https://github.com/usebruno/bruno/assets/18120837/3c735e22-e8cb-4a99-9c45-f8a06bb33463

Dark Mode
**Note:** This looks even nicer with this change #2030 
<img width="795" alt="image" src="https://github.com/usebruno/bruno/assets/18120837/da9ebe46-1ea0-4038-a8fa-78986cf8ac1e">

Light Mode (with hover activated)
<img width="815" alt="image" src="https://github.com/usebruno/bruno/assets/18120837/35e90ee4-14c6-4a4d-a170-0a6247f7e76c">


Screenshot of what a really long document looks like now (peep the scrollbar)
![image](https://github.com/usebruno/bruno/assets/18120837/5e554c9e-3a7b-44ee-89aa-36bb4fda8809)




### Contribution Checklist:

- [x] **The pull request only addresses one issue or adds one feature.**
- [x] **The pull request does not introduce any breaking changes**
- [x] **I have added screenshots or gifs to help explain the change if applicable.**
- [x] **I have read the [contribution guidelines](https://github.com/usebruno/bruno/blob/main/contributing.md).**
- [x] **Create an issue and link to the pull request.**

Note: Keeping the PR small and focused helps make it easier to review and merge. If you have multiple changes you want to make, please consider submitting them as separate pull requests.

### Publishing to New Package Managers

Please see [here](../publishing.md) for more information.

lizziemac committed:
Introduce a reusable MarkDownEditor with improved UX

lizziemac committed:
Make sure onSave is called for button press so we see the toast

lizziemac committed:
Remove unnecessary console.log uses

bpoulaindev commented:
Looks nice but why a quadruple click ?
I submitted a big PR a while ago to introduce Monaco Editor on top of the current one we got (codemirror), i can't give you any date but it will be in beta someday, and will improve the IDE experience for whatever language you choose

lizziemac commented:
Thanks for the comment @bpoulaindev! I changed it to quadruple so that when in the preview, you can select words and paragraphs without going into edit mode.  With that being said, I totally get it if four clicks is overkill and am happy to revert

busy-panda commented:
Hi @lizziemac 

Here are my feedbacks:

- Fixed UI height: The height of the Markdown UI has been adjusted to prevent it from hiding the collection tabs when the documentation is very long. **>> yes it's a good idea**

- Improved save feedback: A toast now pops up when the content is saved by either blurring, hitting cmd+s, or by clicking the Save button. **>> yes it's a good idea**

- Adjusted double-click behavior: Double-clicking no longer takes you into edit mode, making it easier to copy/paste from the docs. Instead, you can quadruple click. This allows users to use the native "double-click" and "triple-click" functionality of text boxes. If this doesn't go over well, reverting is fine. **>> sorry but I don't like the quadruple click to enter edit mode. I can survive if it enters in edit mode when I want to copy/past from the doc**

- Added placeholder text: Placeholder text has been added to guide users to write documentation of their collection! **>> see comment above**

- Enhanced button design: The button has been redesigned and repositioned for a nicer experience.
Hover enables a nice background color **>> Ok but can you add also a cancel button if the user doesn't want to save ? Also I don't like that it saves automatically when you close the tab

lizziemac commented:
Thanks for the feedback! I totally forgot about wanting to cancel editing 😆 I'll try to address your comments soon!  I'm thinking when we leave a tab, give the user a pop up asking to save or cancel

lizziemac committed:
Enable reverting documentation changes with a cancel button

lizziemac committed:
Ensure when user leaves tab while editing, then returns, the UI is still in Edit mode

lizziemac commented:
Hi @busy-panda and @bpoulaindev 

I have addressed all current feedback. Changes made are:
- Double-click to edit re-introduced
- Leaving the tab without saving now replicates behavior of other tabs in the collection.
  - When leaving the tab while editing, then returning to the tab, the doc is still in Edit mode
  - No pop-up currently as I believe that would scope-creep the PR, because it would affect the behavior of the entire collection.
- Cancel button lets you revert changes (image of button below)
    <img width="802" alt="image" src="https://github.com/usebruno/bruno/assets/18120837/8ef1a2ab-f736-45d6-8929-e2c0bd914663">

busy-panda commented:
Hi @lizziemac 

I still have these two points that I think would be great to address:

1) You made the change only in the Collection docs tab, but the same should be done in the request docs tab
2) I don't think it's necessary to change the background color when the user hover with the mouse. Just changing the cursor like you do is enough to indicate that it is editable if he double clicks.

Cheers

lizziemac commented:
Hey @busy-panda !
> 1. You made the change only in the Collection docs tab, but the same should be done in the request docs tab

I was planning to make this a separate PR, but I'll see how much more code it is. I mostly kept them separate because I didn't want to overwhelm with a huge change, especially before getting approval on the design.

> 2. I don't think it's necessary to change the background color when the user hover with the mouse. Just changing the cursor like you do is enough to indicate that it is editable if he double clicks.

I'd like to keep the hover color, I think the change in the cursor style can easily be missed, and the hover color is something other GUIs do as well for this section (I'm thinking Postman specifically)

lizziemac committed:
Add request docs update

lizziemac commented:
marking as draft because introducing requests resulted in some unexpected side effects

lizziemac committed:
code mirror tweaks

lizziemac committed:
Fix bug where calculatedHeight was messing things up

lizziemac committed:
Clean up bottom 'padding' from enabling both x and y overflow scrolling

lizziemac commented:
@busy-panda I've addressed your comment about adding this to the request tab as well, and it is now responsive! Video below with updates. I improved the UI a bit further just because with the response panel things were getting less clean unless I made changes.
Request responsiveness examples
<img width="1624" alt="image" src="https://github.com/usebruno/bruno/assets/18120837/7ca8ea85-fda9-405a-b1ba-54545e48f89b">
<img width="1624" alt="image" src="https://github.com/usebruno/bruno/assets/18120837/4fc8d70e-85b8-48a5-8041-09419363b2a0">

helloanoop commented:
Thanks for working on this @lizziemac !

Some nice ideas here on how to improve the current docs editor experience.
As this PR involves a lot of UX changes, we would need some time to review this and share our feedback.

Given the current backlog of PR's, it will take us some time to get to this.

lizziemac commented:
thanks @helloanoop ! - I can rebase and get the code on the latest after your team is able to review and provide any additional feedback, or ping me if you'd like it sooner

lizziemac commented on packages/bruno-app/src/components/MarkDownEditor/index.jsx:
```diff
@@ -0,0 +1,155 @@
+import MarkdownIt from 'markdown-it';
+import * as MarkdownItReplaceLink from 'markdown-it-replace-link';
```
addressed merged conflicts to work with  https://github.com/usebruno/bruno/pull/2375
